### ISSUE TITLE
linstor-operator-piraeus-cs LinstorController stuck while trying to remove.

### ISSUE THREAD

**Issue by carukc** [piraeusdatastore/piraeus-operator#402]
Hello,

I attempted to install Linstor using the piraeus operator (on Rancher) but something was not quite right in the yaml file and the install failed.

After cancelling the install I attempted to remove the app by uninstalling but that did not work.  I then tried to remove the namespace that I created for the app but that also did not work.

The namespace is showing these two conditions:
* [SomeResourcesRemain] Some resources are remaining: linstorcontrollers.piraeus.linbit.com has 1 resource instances, secrets. has 1 resource instances
* [SomeFinalizersRemain] Some content in the namespace has finalizers remaining: finalizer.linstor-controller.linbit.com in 1 resource instances, piraeus.linbit.com/protect-master-passphrase in 1 resource instances

Rancher has been reporting 'Resource scheduled for deletion' for the last day and I suspect that the secret can not be removed because the controller still exists.

The controller's current status is:
Status:
  Controller Properties:
  Controller Status:
    Node Name:
    Registered On Controller:  false
  Satellite Statuses:
  Errors:
    temporary error: timeout=10000000000, error=failed to contact controller: Get "http://linstor-operator-piraeus-cs.linstor-system.svc:3370/v1/controller/version": dial tcp 10.43.67.84:3370: connect: network is unreachable
    
I cannot get logs because "Error from server (NotFound): pods "LinstorController" not found"

I have tried to force the deletion of the controller using : 
  kubectl delete linstorcontrollers linstor-operator-piraeus-cs -n linstor-system --force

the command starts with the warning:

  Warning: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely.
  linstorcontroller.piraeus.linbit.com "linstor-operator-piraeus-cs" force deleted

but the command simply hangs.

I have also tried patching the controller to remove any finalizers but I get this error:

  kubectl patch linstorcontrollers linstor-operator-piraeus-cs -n linstor-system -p '{"metadata":{"finalizers":[]}}' – type=merge

  application/strategic-merge-patch+json is not supported by piraeus.linbit.com/v1, Kind=LinstorController: the body of the request was in an unknown format - accepted media types include: application/json-patch+json, application/merge-patch+json, application/apply-patch+yaml
  Error from server (NotFound): linstorcontrollers.piraeus.linbit.com "–" not found
  Error from server (NotFound): linstorcontrollers.piraeus.linbit.com "type=merge" not found

I have managed to install successfully in another namespace, however the linstor utility will not work.  It reports:

  2023/02/05 20:15:58 found more than one LinstorController resource:

So I cannot manage the storage cluster.

Is there another way to force the removal of a broken controller?

Thanks

**Comment by carukc** [comment]
I've also tried redeploying on the problematic namespace but kubernetes will not permit me to install while a similar resource is being terminated.

**Comment by carukc** [comment]
OK... took a bit of research but I found a solution and will post here just in case this happens to someone else.

The issue is related to removing the namespace when an install fails.  That's a bad idea because there are finalizers that need to run before the resources can be deleted and if the controller can not be reached then the finalizer can never complete the check that the resource is OK to delete.

This article describes the issue: [The Hidden Dangers of Terminating Namespaces](https://cloud.redhat.com/blog/the-hidden-dangers-of-terminating-namespaces)

To solve the problem first identify the resources that are lingering.  I am not sure how to do that with kubectl but in Rancher you can see it by interrogating the resource that's stuck in the 'Terminating' state in the cluster deployments.

Once you have identified the offending resources, use these commands to tell Kubernetes to ignore the finalizers:

`kubectl patch secret linstor-operator-piraeus-passphrase -n linstor-system -p '{"metadata":{"finalizers":[]}}' --type='merge'`
`kubectl patch LinstorController linstor-operator-piraeus-cs -n linstor-system -p '{"metadata":{"finalizers":[]}}' --type='merge'`

Use the name of the namespace that you created for Linstor rather than linstor-system (which is what I called it when I installed Linstor using the Piraeus Operator).

Basically this removes the finalizers so thta the deletion can take place.  See the [kubectl patch](https://jamesdefabia.github.io/docs/user-guide/kubectl/kubectl_patch/) for more information on how patching works.

Once done... the lingering resources will disappear on their own.

In my attempts to resolve above I got the formatting wrong on the patch command.

I hope that this helps anyone else who stumbles into this trap.
